// graphql-codegen-core/dist/index.js
"use strict";
function __export(m) {
    for (var p in m) if (!exports.hasOwnProperty(p)) exports[p] = m[p];
}
Object.defineProperty(exports, "__esModule", { value: true });
var schema_to_template_context_1 = require("./schema/schema-to-template-context");
exports.schemaToTemplateContext = schema_to_template_context_1.schemaToTemplateContext;
var transform_document_1 = require("./operations/transform-document");
exports.transformDocument = transform_document_1.transformDocument;
exports.transformDocumentsFiles = transform_document_1.transformDocumentsFiles;
var parse_1 = require("./utils/parse");
exports.parse = parse_1.parse;
var to_pascal_case_1 = require("./utils/to-pascal-case");
exports.toPascalCase = to_pascal_case_1.toPascalCase;
__export(require("./types"));
var debugging_1 = require("./debugging");
exports.debugLog = debugging_1.debugLog;
exports.printLogs = debugging_1.printLogs;
exports.resetLogs = debugging_1.resetLogs;
var logger_1 = require("./utils/logger");
exports.getLogger = logger_1.getLogger;
exports.setLogger = logger_1.setLogger;
exports.setSilentLogger = logger_1.setSilentLogger;
exports.useWinstonLogger = logger_1.useWinstonLogger;
