// graphql-toolkit/dist/epoxy/typedefs-mergers/merge-typedefs.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const graphql_1 = require("graphql");
const utils_1 = require("./utils");
const merge_nodes_1 = require("./merge-nodes");
function mergeGraphQLSchemas(types, config) {
    console.info(`
    GraphQL Toolkit/Epoxy 
    Deprecation Notice;
    'mergeGraphQLSchemas' is deprecated and will be removed in the next version.
    Please use 'mergeTypeDefs' instead!
  `);
    return mergeGraphQLTypes(types, config);
}
exports.mergeGraphQLSchemas = mergeGraphQLSchemas;
function mergeTypeDefs(types, config) {
    return {
        kind: 'Document',
        definitions: mergeGraphQLTypes(types, Object.assign({ useSchemaDefinition: true }, config)),
    };
}
exports.mergeTypeDefs = mergeTypeDefs;
function fixSchemaAst(schema) {
    return graphql_1.buildASTSchema(graphql_1.parse(graphql_1.printSchema(schema)));
}
function createSchemaDefinition(def) {
    const schemaRoot = {};
    if (def.query) {
        schemaRoot.query = def.query.toString();
    }
    if (def.mutation) {
        schemaRoot.mutation = def.mutation.toString();
    }
    if (def.subscription) {
        schemaRoot.subscription = def.subscription.toString();
    }
    const fields = Object.keys(schemaRoot)
        .map(rootType => (schemaRoot[rootType] ? `${rootType}: ${schemaRoot[rootType]}` : null))
        .filter(a => a);
    if (fields.length) {
        return `schema { ${fields.join('\n')} }`;
    }
    return undefined;
}
function mergeGraphQLTypes(types, config) {
    const allNodes = types
        .map(type => {
        if (utils_1.isGraphQLSchema(type)) {
            let schema = type;
            let typesMap = type.getTypeMap();
            const validAstNodes = Object.keys(typesMap).filter(key => typesMap[key].astNode);
            if (validAstNodes.length === 0 && Object.keys(typesMap).length > 0) {
                schema = fixSchemaAst(schema);
                typesMap = schema.getTypeMap();
            }
            const schemaDefinition = createSchemaDefinition({
                query: schema.getQueryType(),
                mutation: schema.getMutationType(),
                subscription: schema.getSubscriptionType(),
            });
            const allTypesPrinted = Object.keys(typesMap)
                .map(key => typesMap[key])
                .filter(type => {
                const isPredefinedScalar = type instanceof graphql_1.GraphQLScalarType && graphql_1.isSpecifiedScalarType(type);
                const isIntrospection = graphql_1.isIntrospectionType(type);
                return !isPredefinedScalar && !isIntrospection;
            })
                .map(type => {
                if (type.astNode) {
                    return graphql_1.print(type.extensionASTNodes ? extendDefinition(type) : type.astNode);
                }
                else {
                    return graphql_1.printType(type);
                }
            })
                .filter(e => e);
            const directivesDeclaration = schema
                .getDirectives()
                .map(directive => (directive.astNode ? graphql_1.print(directive.astNode) : null))
                .filter(e => e);
            const printedSchema = [...directivesDeclaration, ...allTypesPrinted, schemaDefinition].join('\n');
            return graphql_1.parse(printedSchema);
        }
        else if (utils_1.isStringTypes(type) || utils_1.isSourceTypes(type)) {
            return graphql_1.parse(type);
        }
        return type;
    })
        .map(ast => ast.definitions)
        .reduce((defs, newDef) => [...defs, ...newDef], []);
    // XXX: right now we don't handle multiple schema definitions
    let schemaDef = allNodes.filter(utils_1.isSchemaDefinition).reduce((def, node) => {
        node.operationTypes
            .filter(op => op.type.name.value)
            .forEach(op => {
            def[op.operation] = op.type.name.value;
        });
        return def;
    }, {
        query: null,
        mutation: null,
        subscription: null,
    });
    const mergedNodes = merge_nodes_1.mergeGraphQLNodes(allNodes);
    const allTypes = Object.keys(mergedNodes);
    if (config && config.useSchemaDefinition) {
        const queryType = schemaDef.query ? schemaDef.query : allTypes.find(t => t === 'Query');
        const mutationType = schemaDef.mutation ? schemaDef.mutation : allTypes.find(t => t === 'Mutation');
        const subscriptionType = schemaDef.subscription ? schemaDef.subscription : allTypes.find(t => t === 'Subscription');
        schemaDef = {
            query: queryType,
            mutation: mutationType,
            subscription: subscriptionType,
        };
    }
    const schemaDefinition = createSchemaDefinition(schemaDef);
    if (!schemaDefinition) {
        return Object.values(mergedNodes);
    }
    return [...Object.values(mergedNodes), graphql_1.parse(schemaDefinition).definitions[0]];
}
exports.mergeGraphQLTypes = mergeGraphQLTypes;
function extendDefinition(type) {
    switch (type.astNode.kind) {
        case 'ObjectTypeDefinition':
            return Object.assign({}, type.astNode, { 
                // add fields from object extension (`extend type Query { newField: String }`)
                fields: type.astNode.fields.concat(type.extensionASTNodes.reduce((fields, node) => fields.concat(node.fields), [])) });
        default:
            return type.astNode;
    }
}
