// graphql-toolkit/dist/loaders/schema/introspection-from-url.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const request_1 = require("request");
const graphql_1 = require("graphql");
const valid_url_1 = require("valid-url");
class IntrospectionFromUrlLoader {
    canHandle(pointerToSchema) {
        return !!valid_url_1.isUri(pointerToSchema);
    }
    handle(url, schemaOptions) {
        let headers = {};
        if (schemaOptions) {
            if (Array.isArray(schemaOptions.headers)) {
                headers = schemaOptions.headers.reduce((prev, v) => (Object.assign({}, prev, v)), {});
            }
            else if (typeof schemaOptions.headers === 'object') {
                headers = schemaOptions.headers;
            }
        }
        let extraHeaders = Object.assign({ Accept: 'application/json', 'Content-Type': 'application/json' }, headers);
        return new Promise((resolve, reject) => {
            request_1.post({
                url: url,
                json: {
                    query: graphql_1.introspectionQuery,
                },
                headers: extraHeaders,
            }, (err, _response, body) => {
                if (err) {
                    reject(err);
                    return;
                }
                const bodyJson = body.data;
                let errorMessage;
                if (body.errors && body.errors.length > 0) {
                    errorMessage = body.errors.map((item) => item.message).join(', ');
                }
                else if (!bodyJson) {
                    errorMessage = body;
                }
                if (errorMessage) {
                    reject('Unable to download schema from remote: ' + errorMessage);
                    return;
                }
                if (!bodyJson.__schema) {
                    throw new Error('Invalid schema provided!');
                }
                resolve(graphql_1.buildClientSchema(bodyJson));
            });
        });
    }
}
exports.IntrospectionFromUrlLoader = IntrospectionFromUrlLoader;
