// graphql-codegen-core/dist/schema/transform-enum.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var debugging_1 = require("../debugging");
var graphql_toolkit_1 = require("graphql-toolkit");
function transformGraphQLEnum(schema, graphqlEnum) {
    debugging_1.debugLog("[transformGraphQLEnum] transformed enum " + graphqlEnum.name);
    var directives = graphql_toolkit_1.getDirectives(schema, graphqlEnum);
    var enumValues = graphqlEnum.getValues().map(function (enumItem) {
        var valueDirectives = graphql_toolkit_1.getDirectives(schema, enumItem);
        return {
            name: enumItem.name,
            description: enumItem.description || '',
            value: enumItem.value,
            directives: valueDirectives,
            usesDirectives: Object.keys(valueDirectives).length > 0
        };
    });
    return {
        name: graphqlEnum.name,
        description: graphqlEnum.description || '',
        values: enumValues,
        directives: directives,
        usesDirectives: Object.keys(directives).length > 0
    };
}
exports.transformGraphQLEnum = transformGraphQLEnum;
