// graphql-toolkit/dist/loaders/schema/schema-from-export.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const fs_1 = require("fs");
const path_1 = require("path");
const isValidPath = require("is-valid-path");
const graphql_1 = require("graphql");
class SchemaFromExport {
    canHandle(pointerToSchema) {
        const fullPath = path_1.isAbsolute(pointerToSchema) ? pointerToSchema : path_1.resolve(process.cwd(), pointerToSchema);
        return isValidPath(pointerToSchema) && fs_1.existsSync(fullPath) && path_1.extname(pointerToSchema) !== '.json';
    }
    async handle(file, _options) {
        const fullPath = path_1.isAbsolute(file) ? file : path_1.resolve(process.cwd(), file);
        if (fs_1.existsSync(fullPath)) {
            const exports = require(fullPath);
            if (exports) {
                let rawExport = exports.default || exports.schema || exports;
                if (rawExport) {
                    let schema = await rawExport;
                    try {
                        const schemaResult = await this.resolveSchema(schema);
                        return schemaResult;
                    }
                    catch (e) {
                        throw new Error('Exported schema must be of type GraphQLSchema, text, AST, or introspection JSON.');
                    }
                }
                else {
                    throw new Error(`Invalid export from export file ${fullPath}: missing default export or 'schema' export!`);
                }
            }
            else {
                throw new Error(`Invalid export from export file ${fullPath}: empty export!`);
            }
        }
        else {
            throw new Error(`Unable to locate introspection from export file: ${fullPath}`);
        }
    }
    isSchemaText(obj) {
        return typeof obj === 'string';
    }
    isWrappedSchemaJson(obj) {
        const json = obj;
        return json.data !== undefined && json.data.__schema !== undefined;
    }
    isSchemaJson(obj) {
        const json = obj;
        return json !== undefined && json.__schema !== undefined;
    }
    isSchemaObject(obj) {
        return obj instanceof graphql_1.GraphQLSchema;
    }
    isSchemaAst(obj) {
        return obj.kind !== undefined;
    }
    isPromise(obj) {
        return !!obj && (typeof obj === 'object' || typeof obj === 'function') && typeof obj.then === 'function';
    }
    async resolveSchema(schema) {
        if (this.isSchemaObject(schema)) {
            return schema;
        }
        else if (this.isSchemaAst(schema)) {
            return graphql_1.buildASTSchema(schema);
        }
        else if (this.isSchemaText(schema)) {
            const ast = graphql_1.parse(schema);
            return graphql_1.buildASTSchema(ast);
        }
        else if (this.isWrappedSchemaJson(schema)) {
            return graphql_1.buildClientSchema(schema.data);
        }
        else if (this.isSchemaJson(schema)) {
            return graphql_1.buildClientSchema(schema);
        }
        else {
            throw new Error('Unexpected schema type provided!');
        }
    }
}
exports.SchemaFromExport = SchemaFromExport;
